File: srui-core/src/events.rs
//! Output events — the core's entire output surface.
//!
//! The core never speaks. It emits structured events; readers (speech,
//! braille, UIA, test harnesses) consume them and render as they see fit.
//! The `speech` module provides the reference rendering.

use crate::tree::NodeId;
use crate::types::WidgetLabel;

/// A single item in the output stream.
#[derive(Debug, Clone, PartialEq)]
pub enum OutputEvent {
    /// What the user should perceive. Consumed by readers.
    Accessibility(AccessibilityEvent),
    /// What the program should react to. Consumed by application code.
    Widget(WidgetEvent),
    /// A ticker's interval elapsed (see `Ui::add_ticker`). Fires at most
    /// once per `set_now` call per ticker, so its resolution is the
    /// host's clock cadence.
    Tick { ticker: u64 },
}

/// Semantic events for the program: user intent, not perception.
#[derive(Debug, Clone, PartialEq)]
pub enum WidgetEvent {
    /// The node was activated (Enter on a button, or the layer's
    /// primary/cancel widget was triggered).
    Activated { node: NodeId },
    /// Secondary activation (Shift+Enter).
    SecondaryActivated { node: NodeId },
    /// A checkbox was toggled. `checked` is the new value.
    Toggled { node: NodeId, checked: bool },
    /// The node's widget state changed (text edited, selection moved,
    /// slider adjusted).
    Changed { node: NodeId },
}

/// Structured payload describing what the user should perceive. Each
/// variant carries the data its non-speech consumers (braille, UIA,
/// sonification) will need; the label snapshot is taken at emission time
/// so readers never race the tree.
#[derive(Debug, Clone, PartialEq)]
pub enum AccessibilityEvent {
    /// Focus moved to a different widget.
    Focused {
        node: NodeId,
        label: WidgetLabel,
        /// Names of Label-role siblings preceding the focused widget in
        /// child order. Empty on ordinary focus moves; populated when the
        /// host asks for a context re-announcement after a view
        /// transition.
        context_labels: Vec<String>,
    },

    /// Text was inserted or deleted in an editor. Deliberately carries
    /// only the perceptual content (what the user hears), not document
    /// text — readers that need the line or surrounding text query the
    /// editor through the node.
    Typing {
        node: NodeId,
        /// Single grapheme for `Insert`/`Delete`; empty for `DeleteWord`.
        grapheme: String,
        /// `Some(word)` when an insert just crossed a word boundary or
        /// when `kind` is `DeleteWord`.
        last_word: Option<String>,
        kind: TypingKind,
    },

    /// Cursor moved within text without an edit. Carries perceptual
    /// content only; query the editor for document text.
    TextNav {
        node: NodeId,
        /// Grapheme at the post-move cursor position.
        grapheme_at_cursor: String,
        /// Spoken context for the landing position at this granularity —
        /// the expanded character, the word, or the line. Composed by the
        /// widget because it depends on cursor-adjacent state.
        context: String,
        granularity: NavGranularity,
        /// `Some` when nav was attempted past an edge and the cursor
        /// did not move.
        boundary: Option<Boundary>,
    },

    /// Selection extended, contracted, cleared, or set-all.
    Selection {
        node: NodeId,
        /// The text added to or removed from the selection (or
        /// "{n} characters" for large selections).
        delta: String,
        kind: SelectionKind,
    },

    /// Discrete-value list/grid widget changed selection.
    ItemNav {
        node: NodeId,
        item: String,
        /// `(zero_based_index, total)`. `None` when the widget has no
        /// indexable concept.
        position: Option<(usize, usize)>,
        boundary: Option<Boundary>,
    },

    /// Tab control moved to a different tab. No `boundary` because tabs
    /// are circular.
    TabChange {
        node: NodeId,
        tab_name: String,
        position: (usize, usize),
    },

    /// Slider value changed.
    SliderChange {
        node: NodeId,
        value: i32,
        unit: String,
    },

    /// Filter / typeahead query changed; results recomputed.
    Filter {
        node: NodeId,
        query: String,
        first_result: Option<String>,
        result_count: usize,
    },

    /// Clipboard operation completed.
    Clipboard { node: NodeId, op: ClipboardOp },

    /// Free-form announcement — escape hatch for anything without a
    /// structured variant ("Nothing to delete", "no results", "blank").
    Announce { text: String },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TypingKind {
    Insert,
    Delete,
    DeleteWord,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SelectionKind {
    Selected,
    Unselected,
    Cleared,
    All,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Boundary {
    Top,
    Bottom,
    Left,
    Right,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ClipboardOp {
    Copy,
    Cut,
    Paste,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum NavGranularity {
    Char,
    Word,
    LineEdge,
    TextEdge,
    LineUp,
    LineDown,
}

/// Apply coalescing rules to a drained batch.
///
/// State-describing accessibility events (`Focused`, `Selection`,
/// `ItemNav`, `TabChange`, `SliderChange`, `Filter`) keep only the last
/// occurrence of their kind — an intermediate focus move is discarded in
/// favor of the settled one. Action events (`Typing`, `TextNav`,
/// `Clipboard`, `Announce`) and all widget events keep emission order.
pub fn coalesce(events: Vec<OutputEvent>) -> Vec<OutputEvent> {
    use std::mem::discriminant;

    fn state_event(ev: &OutputEvent) -> Option<std::mem::Discriminant<AccessibilityEvent>> {
        match ev {
            OutputEvent::Accessibility(a) => match a {
                AccessibilityEvent::Focused { .. }
                | AccessibilityEvent::Selection { .. }
                | AccessibilityEvent::ItemNav { .. }
                | AccessibilityEvent::TabChange { .. }
                | AccessibilityEvent::SliderChange { .. }
                | AccessibilityEvent::Filter { .. } => Some(discriminant(a)),
                _ => None,
            },
            OutputEvent::Widget(_) | OutputEvent::Tick { .. } => None,
        }
    }

    // Pass 1: last occurrence index per state-event kind.
    let mut last_idx: Vec<(std::mem::Discriminant<AccessibilityEvent>, usize)> = Vec::new();
    for (i, ev) in events.iter().enumerate() {
        if let Some(d) = state_event(ev) {
            if let Some(slot) = last_idx.iter_mut().find(|(dd, _)| *dd == d) {
                slot.1 = i;
            } else {
                last_idx.push((d, i));
            }
        }
    }

    // Pass 2: drop state events that aren't the last of their kind.
    events
        .into_iter()
        .enumerate()
        .filter(|(i, ev)| match state_event(ev) {
            None => true,
            Some(d) => last_idx
                .iter()
                .find(|(dd, _)| *dd == d)
                .map(|(_, last)| *last == *i)
                .unwrap_or(true),
        })
        .map(|(_, ev)| ev)
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tree::Tree;
    use crate::types::{Role, WidgetLabel};

    fn focused_event(tree: &mut Tree, name: &str) -> OutputEvent {
        let id = tree.insert(None, usize::MAX, WidgetLabel::new(name, Role::Button));
        OutputEvent::Accessibility(AccessibilityEvent::Focused {
            node: id,
            label: WidgetLabel::new(name, Role::Button),
            context_labels: Vec::new(),
        })
    }

    #[test]
    fn coalesce_keeps_last_focused() {
        let mut tree = Tree::new();
        let a = focused_event(&mut tree, "A");
        let b = focused_event(&mut tree, "B");
        let out = coalesce(vec![a, b.clone()]);
        assert_eq!(out, vec![b]);
    }

    #[test]
    fn coalesce_keeps_all_announces() {
        let one = OutputEvent::Accessibility(AccessibilityEvent::Announce { text: "one".into() });
        let two = OutputEvent::Accessibility(AccessibilityEvent::Announce { text: "two".into() });
        let out = coalesce(vec![one.clone(), two.clone()]);
        assert_eq!(out, vec![one, two]);
    }

    #[test]
    fn coalesce_preserves_widget_events() {
        let mut tree = Tree::new();
        let a = focused_event(&mut tree, "A");
        let b = focused_event(&mut tree, "B");
        let id = tree.insert(None, usize::MAX, WidgetLabel::new("X", Role::Button));
        let w = OutputEvent::Widget(WidgetEvent::Activated { node: id });
        let out = coalesce(vec![a, w.clone(), b.clone()]);
        assert_eq!(out, vec![w, b]);
    }

    #[test]
    fn coalesce_is_per_kind_not_global() {
        let mut tree = Tree::new();
        let f = focused_event(&mut tree, "A");
        let id = tree.insert(None, usize::MAX, WidgetLabel::new("L", Role::ListBox));
        let item = OutputEvent::Accessibility(AccessibilityEvent::ItemNav {
            node: id,
            item: "first".into(),
            position: Some((0, 3)),
            boundary: None,
        });
        // Focused and ItemNav are different kinds — both survive.
        let out = coalesce(vec![f.clone(), item.clone()]);
        assert_eq!(out, vec![f, item]);
    }
}

File: Srui.Net/Events.cs
namespace Srui;

/// <summary>Widget states — conditions the user cannot directly change.</summary>
[Flags]
public enum WidgetStates : uint
{
    None = 0,
    Disabled = 1 << 0,
    Required = 1 << 1,
    Warning = 1 << 2,
    Hidden = 1 << 3,
    /// <summary>The widget carries <see cref="Widget.KeyHelp"/> text;
    /// spoken "with help", read on F1.</summary>
    WithHelp = 1 << 4,
}

/// <summary>What a Typing event describes.</summary>
public enum TypingKind
{
    Insert,
    Delete,
    DeleteWord,
}

/// <summary>What a Selection event describes.</summary>
public enum SelectionKind
{
    Selected,
    Unselected,
    Cleared,
    All,
}

/// <summary>An edge the user ran into (or landed on) while navigating.</summary>
public enum Boundary
{
    Top,
    Bottom,
    Left,
    Right,
}

public enum ClipboardOp
{
    Copy,
    Cut,
    Paste,
}

/// <summary>What an EditNoop event reports — the editor input that had
/// nothing to act on.</summary>
public enum EditNoopKind
{
    NoText,
    NothingToSelect,
    NothingToDelete,
    SelectedToTop,
    SelectedToBottom,
}

/// <summary>The label property a LabelChange event names.</summary>
public enum LabelPart
{
    Name,
    Role,
    Value,
    Description,
}

/// <summary>Why a Focused event was emitted — the provenance readers
/// filter on (an earcon reader clicks on user movement but not on a
/// dialog opening; a verbosity policy may say more on recovery).</summary>
public enum FocusCause
{
    /// <summary>The user moved: the tab ring or hierarchy navigation.</summary>
    UserNavigation,
    /// <summary>A widget shortcut jumped focus (mnemonics included).</summary>
    Shortcut,
    /// <summary>The program moved focus: Widget.Focus() or EnsureFocus.</summary>
    Programmatic,
    /// <summary>Focus recovered after the focused node was removed or
    /// became unreachable.</summary>
    Recovery,
    /// <summary>A layer was popped and the previous layer's focus
    /// restored.</summary>
    LayerRestore,
    /// <summary>Focus did not move: a re-announcement of the focused
    /// widget (speak-focus, or the with-context announcement after a
    /// view transition).</summary>
    Reannounce,
}

/// <summary>Granularity of a text-cursor move.</summary>
public enum NavGranularity
{
    Char,
    Word,
    LineEdge,
    TextEdge,
    LineUp,
    LineDown,
}

/// <summary>A snapshot of the six semantic properties a screen reader
/// announces, in NVDA order: name, role, value, states, description,
/// shortcut. Role is the spoken role text (empty for role-less widgets);
/// StateText is dynamic state spoken before the flag states (e.g. "no
/// filter"); Shortcuts lists every combo attached to the widget, first
/// added first (announcements speak the first). Taken at emission time,
/// so readers never race the tree.</summary>
public sealed record WidgetInfo(
    string? Name,
    string Role,
    string Value,
    string StateText,
    WidgetStates States,
    string Description,
    IReadOnlyList<KeyCombo> Shortcuts);

/// <summary>Structured payload describing what the user should perceive.
/// Readers consume these; the reference speech rendering lives in
/// <see cref="SpeechRenderer"/>. Every variant carries the widget it
/// concerns; on Announce the source is optional (null for app-level
/// announcements).</summary>
public abstract record AccessibilityEvent
{
    /// <summary>Focus moved to (or was re-announced on) a widget. Info is
    /// the golden-six snapshot at emission. ContextLabels holds the names
    /// of Label siblings preceding the widget in child order — empty on
    /// ordinary focus moves, populated for context re-announcements after
    /// a view transition (dialog openings). Cause is the provenance of
    /// the move (user navigation, shortcut jump, programmatic, recovery,
    /// layer restore, re-announcement), so readers can treat user
    /// movement differently from focus that came to the user.</summary>
    public sealed record Focused(
        Widget Widget, WidgetInfo Info, IReadOnlyList<string> ContextLabels, FocusCause Cause)
        : AccessibilityEvent;

    /// <summary>Text was inserted or deleted in an editor. Carries only
    /// the perceptual content (what the user hears), not document text.
    /// Grapheme is a single grapheme for Insert/Delete and empty for
    /// DeleteWord; LastWord is set when an insert just crossed a word
    /// boundary or when the kind is DeleteWord.</summary>
    public sealed record Typing(Widget Widget, string Grapheme, string? LastWord, TypingKind Kind)
        : AccessibilityEvent;

    /// <summary>Cursor moved within text without an edit. Context is the
    /// spoken content for the landing position at this granularity — the
    /// expanded character, the word, or the line. BoundaryHit is set when
    /// nav was attempted past an edge and the cursor did not move.</summary>
    public sealed record TextNav(
        Widget Widget, string GraphemeAtCursor, string Context,
        NavGranularity Granularity, Boundary? BoundaryHit) : AccessibilityEvent;

    /// <summary>Selection extended, contracted, cleared, or set-all.
    /// Delta is the text added to or removed from the selection (or
    /// "{n} characters" for large selections).</summary>
    public sealed record Selection(Widget Widget, string Delta, SelectionKind Kind)
        : AccessibilityEvent;

    /// <summary>Discrete-value list/grid widget changed selection.
    /// Position is (zero-based index, total), or null when the widget has
    /// no indexable concept.</summary>
    public sealed record ItemNav(
        Widget Widget, string Item, (int Index, int Total)? Position, Boundary? BoundaryHit)
        : AccessibilityEvent;

    /// <summary>Tab control moved to a different tab. No boundary because
    /// tabs are circular.</summary>
    public sealed record TabChange(Widget Widget, string TabName, (int Index, int Total) Position)
        : AccessibilityEvent;

    /// <summary>Slider value changed.</summary>
    public sealed record SliderChange(Widget Widget, int Value, string Unit) : AccessibilityEvent;

    /// <summary>Check box toggled — by the user, or programmatically
    /// while focused. The reference rendering speaks the new state alone
    /// ("checked", "not checked").</summary>
    public sealed record Toggle(Widget Widget, bool Checked) : AccessibilityEvent;

    /// <summary>Filter / typeahead query changed; results recomputed.</summary>
    public sealed record Filter(Widget Widget, string Query, string? FirstResult, int ResultCount)
        : AccessibilityEvent;

    /// <summary>Clipboard operation completed.</summary>
    public sealed record Clipboard(Widget Widget, ClipboardOp Op) : AccessibilityEvent;

    /// <summary>An editor input with nothing to act on: navigation or
    /// selection in an empty editor, a delete with nothing beside the
    /// cursor, or a selection already pinned at a text edge. Context is
    /// the spoken content at the cursor for the pinned-selection kinds,
    /// null otherwise.</summary>
    public sealed record EditNoop(Widget Widget, EditNoopKind Kind, string? Context = null)
        : AccessibilityEvent;

    /// <summary>A label property of the focused widget changed
    /// programmatically. Text is the property's new spoken text; the
    /// reference rendering speaks it alone — the delta, never a full
    /// re-announcement. Label changes on unfocused widgets are silent
    /// and produce no event.</summary>
    public sealed record LabelChange(Widget Widget, LabelPart Part, string Text)
        : AccessibilityEvent;

    /// <summary>A state flag of the focused widget was set or cleared
    /// programmatically ("disabled", "required", "warning cleared").
    /// Hidden never appears here: hiding the focused widget reads as the
    /// focus recovery it causes.</summary>
    public sealed record StateChange(Widget Widget, WidgetStates State, bool On)
        : AccessibilityEvent;

    /// <summary>Free-form announcement — the escape hatch for anything
    /// without a structured variant. Source is the widget that emitted
    /// it, or null for app-level announcements.</summary>
    public sealed record Announce(string Text, Widget? Source = null) : AccessibilityEvent;
}

/// <summary>A consumer of accessibility events: self-voicing speech,
/// braille, a platform accessibility bridge, a test recorder. Readers are
/// attached with <see cref="SruiApp.AddReader"/> and receive every
/// accessibility event the drain delivers, in order.</summary>
public interface IReader
{
    void OnEvent(AccessibilityEvent e);

    /// <summary>The user acted (a key went down, or the app demanded
    /// urgency): whatever is being presented is stale. The speech reader
    /// silences here; readers with no notion of interruption ignore it.</summary>
    void OnInterrupt()
    {
    }
}

/// <summary>One pumped host event.</summary>
public abstract record HostEvent
{
    public sealed record Quit : HostEvent;

    /// <summary>A physical key went down; readers are interrupted before
    /// the corresponding Input is handled.</summary>
    public sealed record KeyDown : HostEvent;

    /// <summary>A clean Alt tap (commonly bound to a menu/palette).</summary>
    public sealed record AltTap : HostEvent;

    public sealed record Input(InputEvent Event) : HostEvent;

    /// <summary>A physical key transition (press, repeat, or release),
    /// parallel to and independent of the Input stream.</summary>
    public sealed record Key(KeyInput Event) : HostEvent;

    /// <summary>The window lost keyboard focus. Held-key releases will
    /// not arrive; zero any held-key state.</summary>
    public sealed record FocusLost : HostEvent;
}
